### DISCUSSION TITLE
Incorrect HTTP Verb for route | Wheels.RouteNotFound

### DISCUSSION THREAD

**Discussion by BrevardCounty-Wade** [wheels-dev/wheels#1225] [Q&A] [answered]
Hello,

I am having some trouble with updating a ColdFusion server to a Lucee + CFWheels setup. The issue was not encountered when running the project on Lucee without CFWheels. However, CFWheels has been rather helpful in its development, and I'd like to keep its inclusion as a framework.

Most of the web app transferred from that pure Lucee version with Tomcat URL rewriting. However, when I tried to include CFWheels in a separate new version, only one operation did not perform as expected. It involves a form with a jQuery 'post()' method using JSON for form elements to be arguments in a CFC file called 'auth.cfc'. I did include Tomcat URL rewriting, and I have a temporary setup for wildcard routing with GET and POST methods. I also updated the /config/routes.cfm to include the following line:

> .resource(name="auth", controller="auth")

After updating the URL path to match the planned changes and accounting for the suggested tomcat URL rewriting in documentation I had done earlier, the error updated to a Wheels.RouteNotFound error. I have no other alterations nor config settings that would interfere with the POST method. I have rewritten the jQuery operation to be a CFML condition using CGI scope for the time being, and I did a 'cfinvoke' function to the CFC file with the necessary arguments and values. I have noted the JSON is returned, but I am now landing on the /rewrite.cfm view with the following message:

> **Wheels.RouteNotFound**
> Incorrect HTTP Verb for route
>
> **Suggested action**
> The path does not allow POST requests, only GET requests. Ensure you are using the correct HTTP Verb and that your config/routes.cfm file is configured correctly.

I have performed the following actions, but the routing error persists.

1. Compared this POST method using a CGI if statement with other successful submission events to try to mimic them.
2. In routes.cfm, I included the following statement for the testing environment only: .wildcard(methods="get,post")
3. Tried to have the form handling CGI if statement set the 'cflocation' event to be the full URL rather than the shorten route.
4. Commented out the 'cflocation' event. Created a CFWheels post() route as the intended next page and updated the 'cflocation' to a #redirectTo(name="loginSuccess")# : .post(name="loginSuccess", to="search##name", pattern="/search/name") 

I have been viewing and reading the [CFWheels Routing documentation](https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/routing), but I am at a lost at this point. Could someone help point me in the right direction? Is there a mistake I'm making in debugging this?

**Comment by bpamiri** [comment, marked as answer]
Sorry to hear you're having issues getting routes working. Let's see if we can figure out what is going on. Have you tried reviewing the routes in your app to see if you are call the correct route? The easiest way to do this is to use the /wheels/routes URL to see the routing table.

Can you post your routing table here and the code you are using to call the action.

**Reply by BrevardCounty-Wade** [reply]
Sure, let me know if I left anything out.

**Routes**
Name | Method | Pattern | Controller | Action
-- | -- | -- | -- | --
auth | post | /auth.[format] | auth | create
auth | post | /auth | auth | create
newAuth | get | /auth/new.[format] | auth | new
newAuth | get | /auth/new | auth | new
editAuth | get | /auth/edit.[format] | auth | edit
editAuth | get | /auth/edit | auth | edit
auth | get | /auth.[format] | auth | show
auth | get | /auth | auth | show
auth | patch | /auth.[format] | auth | update
auth | patch | /auth | auth | update
auth | put | /auth.[format] | auth | update
auth | put | /auth | auth | update
auth | delete | /auth.[format] | auth | delete
auth | delete | /auth | auth | delete
loginSuccess | post | /search/name | search | name
wildcard | get | /[controller]/[action] |   |  
wildcard | get | /[controller] |   | index
wildcard | post | /[controller]/[action] |   |  
wildcard | post | /[controller] |   | index
root | get | / | landing | disclaimer

As for the code routing to it, the original jQuery process has been temporarily replaced with a CGI scope if statement below (written after the cfform in the view file).

```coldfusion
<cfif CGI.Request_Method EQ "POST" OR IsDefined("FORM.submit")>
 ...
      <cfcase value="success"> <!--- JSON value --->
              <!--- <cflocation  url="/search/name"> --->
 	         #redirectTo(name="loginSuccess")#
       </cfcase>
 ...
 </cfif>
```

**Reply by bpamiri** [reply]
Can you post the complete error you're getting back as well.

**Reply by BrevardCounty-Wade** [reply]
Sure thing!

> **Wheels.RouteNotFound**
Incorrect HTTP Verb for route
>
> **Suggested action**
The path does not allow POST requests, only GET requests. Ensure you are using the correct HTTP Verb and that your config/routes.cfm file is configured correctly.
>
> **Tag context**
> Error thrown on line 133 in wheels\dispatch\functions.cfm
> - called from line 223 in wheels\dispatch\functions.cfm
> - called from line 178 in wheels\dispatch\functions.cfm
> - called from line 6 in wheels\index.cfm
> - called from line 2 in rewrite.cfm
> - called from line 4 in wheels\events\onrequest.cfm

**Reply by bpamiri** [reply]
What I think might be happening is that your redirectTo(name="loginSuccess") call is really doing a GET to that URL and your routing table only allows POST for that route. But I want to make sure by seeing the error you are getting.

**Reply by neokoenig** [reply]
May not be related but I don't think you can redirect to a POST route...
`loginSuccess` is `POST`, but you're doing `redirectTo()`, which I think will just do a 301/302 redirect, so it will probably match the wildcard route underneath. just FYI.

**Reply by bpamiri** [reply]
Ya, that is what I thought. So this is what is happening. You have a FORM on your page which is posting to an action. That is the actual POST call. Then in the action of that POST call you are redirecting to another action which does a GET redirect. Well, technically it sends back a 302 HTTP response and the browser loads the new location but at the end of the day it does a GET to the loginSuccess route that only allows a POST.

To get rid of your error, you could change the loginSuccess route to accept a GET and that should resolve your error. Not sure the flow makes sense to me but that should take care of your error.

**Reply by neokoenig** [reply]
Yeah wheels doesn't support 307 requests without doing it manually

**Reply by bpamiri** [reply]
Tom is exactly right. But more succinct and a faster typer. ;-)

**Reply by bpamiri** [reply]
Actually looking at the error message I think the error is being generated by the Form call not the redirectTo call. Which is still a problem as mentioned above. Can you post the code in your Form call to see where it is sending the initial Post request.

**Reply by BrevardCounty-Wade** [reply]
Sure.

```coldfusion
<cfform id="loginForm" ... method="POST">
...
	<div class="submissionOptions">
		#authenticityTokenField()#
		<input id="loginSubmit" type="submit" value="Login" name="submit"> 
	</div>
</cfform>

<cfif CGI.Request_Method EQ "POST" OR IsDefined("FORM.submit")>
 ...

**Reply by bpamiri** [reply]
Does your CFFORM actually have an action attribute? or is it left off thus forcing the call to go to the same URL as the page you are on.

When using a HTML form or in this case a CFFORM instead of trying to hard code the URL to call I will typically make use of the urlFor() function so I can still use named routes and let CFWheels figure out the URL to send the call to. So I would use 
```
<form action="#urlFor(route='loginSuccess')#" method="post">
  #authenticityTokenField()#
</form>
```

From the code fragment you posted it would seem like your page is configured like old school form on a page, posting to itself, and an if block on the page to determine if the call is the initial form display call or the secondary form action call. This isn't really necessary when using CFWheels. You can have an action that displays the form which would submit to another action to process the form. You may want to step back and look at what you are trying to achieve and see if there is a better way to accomplish it with CFWheels. If you are trying to use CFForm to make an Ajax call for user interface or users ease of use basis, I would recommend checking out HTMX (htmx.org) which allows for much more modern user interfaces with out the need for jQuery, CFForms, or hand written JavaScript. But that is probably a bigger lift that what you are trying to achieve here.

For this issue, I would try to break it down to its individual call and responses. You have a call that displays the form, then a call to submit the form, then a call to send the user to the success page. Work on each of those individually in that order to see where the error is coming in. Based on what I'm seeing your error is being generated by the CFFORM posting the results of the data entry to a route that only accepts a GET HTTP verb.

**Reply by BrevardCounty-Wade** [reply]
Ah, that did it!

No, I did not include an action attribute to the cfform. I have a onsubmit attribute defined for the jQuery portion, but I forgot to add in an action attribute to account for the accomodations I was trying to make. Once I added that exact action, it submitted the form and loaded the intended page.

Thank you!
